lib/Sema/ConstraintSystem.cpp

```diff
@@ -4448,3 +4425,17 @@ void ConstraintSystem::maybeProduceFallbackDiagnostic(
 
   ctx.Diags.diagnose(target.getLoc(), diag::failed_to_produce_diagnostic);
 }
+
+Type ConstraintSystem::replaceDependentTypes(Type origType) {
```

Comment:
I feel like this is getting called in the wrong places. We should replace DependentMemberTypes with TypeVariableTypes in the same places where we replace GenericTypeParameterTypes with TypeVariableTypes, ie `openGeneric()` and so on. The constraint system should never see types that contain DependentMemberTypes just like it never sees types that contain GenericTypeParamTypes.

Reply:
Tha'ts what I thought as well but in reality after I have added asserts to `getFixedTypeRecursive` and `simplifyTypeImpl` it turned up that conditional conformance requirements has dependent member types and `getSuperclass()` returns type with dependent member types in generic arguments, these types aren't opened. This changes serve more to show the approach so I'm still in search for a right solution how to deal with that...

Reply:
Can you share examples that demonstrate those problems?

Reply:
Sure! There are a multiple cases in stdlib that I found, I'll try to reduce.

Reply:
Thanks! I'm not sure about conditional requirements, but certainly adding constraints with generic parameters in them sounds like an oversight... we might be doing something wrong with those calls to getSuperclass().

Reply:
Indeed, I was surprised that we end up with dependent members there but I think it's because of getContextualSubstitutions since base of the dependent member type is a type variable. I'm going to finish something up first and come back to this.

Reply:
I see, if the base is a dependent member variable, I guess you'll still need your special operation for opening those up. The other alternative is that we can teach Type::subst() how to do that.

Reply:
Yes, that's what I was thinking actually - have a version of `getSupertype` and `conformsToProtocol` which accepts a callback that's going to property handle dependent member types, WDYT?